src/Symfony/Component/Lock/Store/ZookeeperStore.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\LockReleasingException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\StoreInterface;
+
+/**
+ * ZookeeperStore is a StoreInterface implementation using Zookeeper as store engine.
+ *
+ * @author Ganesh Chandrasekaran <gchandrasekaran@wayfair.com>
+ */
+class ZookeeperStore implements StoreInterface
+{
+    /**
+     * @var \Zookeeper
+     */
+    private $zookeeper;
+
+    /**
+     * Locker constructor.
+     *
+     * @param \Zookeeper $zookeeper Zookeeper Instance
+     */
+    public function __construct(\Zookeeper $zookeeper)
+    {
+        $this->zookeeper = $zookeeper;
+    }
+
+    /**
+     * Stores the resource if it's not locked by someone else.
+     *
+     * @param Key $key the Key that represent the lock resource
+     *
+     * @throws LockConflictedException
+     * @throws LockAcquiringException
+     */
+    public function save(Key $key)
+    {
+        $resource = $this->getKeyResource($key);
+
+        // This makes sure the node dies at the end of a client session, thus releasing the lock to avoid starvation
+        $flags = \Zookeeper::EPHEMERAL;
+
+        try {
+            $token = $this->getToken($key);
+
+            if ($this->exists($key)) {
+                if ($token === $this->zookeeper->get($resource)) {
+                    return;
+                }
+            }
+
+            $this->createNewLock($resource, $token, array(), $flags);
+        } catch (LockConflictedException $ex) {
+            throw $ex;
+        } catch (\Throwable $throwable) {
+            throw new LockAcquiringException($throwable);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockReleasingException
+     */
+    public function delete(Key $key)
+    {
+        $resource = $this->getKeyResource($key);
+        try {
+            if ($this->exists($key)) {
+                $this->zookeeper->delete($resource);
+            }
+        } catch (\Throwable $throwable) {
+            // For Zookeeper Ephemeral Nodes, the node will be deleted upon session death. But, if we want to unlock
```

Comment:
Since this store doesn't support expiring locks, ephemeral nodes is the way to go to clean up the locks in case of any failures such as fatal errors etc. The lock will be released successfully in any case. The only case when delete fails is, if a node has children nodes. For example if you have `baseNode/1234/chilNode` and you try to delete `baseNode/1234` the delete would fail. In our case, since we always use 1 node and that too an ephemeral node, this will ideally never fail. By design, ephemeral node cannot have a children node.